### ISSUE TITLE
SQL Server 2008 SELECT DISTINCT .. LIMIT .. OFFSET emulation does not keep ORDER BY semantics in complex queries

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#5068]
The following query returns a wrong result:

```java
assertEquals(asList(4, 3),
create().selectDistinct(
            TBook_ID(),
            TBook_AUTHOR_ID().mul(inline(2)),
            lower(TBook_TITLE()))
        .from(TBook())
        .leftJoin(TAuthor())
            .on(TBook_AUTHOR_ID().eq(TAuthor_ID()))
        .leftJoin(TBookToBookStore())
            .on(TBook_ID().eq(TBookToBookStore_BOOK_ID()))
        .leftJoin(TBookStore())
            .on(TBookStore_NAME().eq(TBookToBookStore_BOOK_STORE_NAME()))

        // Avoid fetching 1984 as ordering differs between DBs
        .where(TBook_ID().between(2).and(4))
        .orderBy(
            TBook_AUTHOR_ID().mul(inline(2)).asc(),
            lower(TBook_TITLE()).asc())

        // Force Sybase to simulate TOP .. START AT
        .limit(inline(2))
        .offset(1)
        .fetch(TBook_ID()));
```

It is generated by jOOQ for SQL Server 2008 as such:

```sql
select 
  [v0] [ID], 
  [v1] [*], 
  [v2] [lower]
from (
  select distinct top 100 percent 
    [test].[dbo].[t_book].[ID] [v0], 
    ([test].[dbo].[t_book].[AUTHOR_ID] * 2) [v1], 
    lower([test].[dbo].[t_book].[TITLE]) [v2], 
    dense_rank() over (order by 
      ([test].[dbo].[t_book].[AUTHOR_ID] * 2) asc, 
      lower([test].[dbo].[t_book].[TITLE]) asc, 
      [test].[dbo].[t_book].[ID] asc, 
      ([test].[dbo].[t_book].[AUTHOR_ID] * 2) asc, 
      lower([test].[dbo].[t_book].[TITLE]) asc
    ) [rn]
  from [test].[dbo].[t_book]
    left outer join [test].[dbo].[t_author]
    on [test].[dbo].[t_book].[AUTHOR_ID] = [test].[dbo].[t_author].[ID]
    left outer join [test].[dbo].[t_book_to_book_store]
    on [test].[dbo].[t_book].[ID] = [test].[dbo].[t_book_to_book_store].[BOOK_ID]
    left outer join [test].[dbo].[t_book_store]
    on [test].[dbo].[t_book_store].[NAME] = [test].[dbo].[t_book_to_book_store].[BOOK_STORE_NAME]
  where [test].[dbo].[t_book].[ID] between 2 and 4
  order by 
    [v1] asc, 
    [v2] asc
  
) [x]
where [rn] > 1
and [rn] <= (1 + 2)
```

The problem here is that the `ORDER BY` clause is not enforced in the outer most query. This is often OK, specifically if there is no `DISTINCT` clause involved, SQL Server will implement stable ordering in the subquery and maintain the ordering also in the top level select.

But it is wrong to rely on this behaviour.

Specifically in the case of `DISTINCT`, it is safe to assume that all columns from the `ORDER BY` clause are also present in the `SELECT` clause, so we can move the `ORDER BY` clause to the top level select, safely.

**Comment by lukaseder** [comment]
The `ORDER BY` clause cannot be repeated as it is quite possible that it uses expressions that are out of scope for the outer-most query, e.g. fully qualified column names, correlated subqueries, etc.

What can be guaranteed, though is ordering by `[rn]`. The above query can be fixed as such:

```sql
select 
  [v0] [ID], 
  [v1] [*], 
  [v2] [lower]
from (
  select distinct top 100 percent 
    [test].[dbo].[t_book].[ID] [v0], 
    ([test].[dbo].[t_book].[AUTHOR_ID] * 2) [v1], 
    lower([test].[dbo].[t_book].[TITLE]) [v2], 
    dense_rank() over (order by 
      ([test].[dbo].[t_book].[AUTHOR_ID] * 2) asc, 
      lower([test].[dbo].[t_book].[TITLE]) asc, 
      [test].[dbo].[t_book].[ID] asc, 
      ([test].[dbo].[t_book].[AUTHOR_ID] * 2) asc, 
      lower([test].[dbo].[t_book].[TITLE]) asc
    ) [rn]
  from [test].[dbo].[t_book]
    left outer join [test].[dbo].[t_author]
    on [test].[dbo].[t_book].[AUTHOR_ID] = [test].[dbo].[t_author].[ID]
    left outer join [test].[dbo].[t_book_to_book_store]
    on [test].[dbo].[t_book].[ID] = [test].[dbo].[t_book_to_book_store].[BOOK_ID]
    left outer join [test].[dbo].[t_book_store]
    on [test].[dbo].[t_book_store].[NAME] = [test].[dbo].[t_book_to_book_store].[BOOK_STORE_NAME]
  where [test].[dbo].[t_book].[ID] between 2 and 4
  order by 
    [v1] asc, 
    [v2] asc

) [x]
where [rn] > 1
and [rn] <= (1 + 2)
order by [rn]
```

**Comment by lukaseder** [comment]
The fix will be merged to 3.7.3 (#5071). Merging it to earlier minor releases seems rather risky, as a lot of things have changed with this emulation in 3.7
